tests/python_package_test/test_sklearn.py

```diff
@@ -2048,3 +2049,101 @@ def test_classifier_fit_detects_classes_every_time():
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_set=[(X_test, y_test)], eval_y=y_test)
+
+    msg = "You must specify eval_X and eval_y, not just one of them."
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_X=X_test)
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_y=y_test)
+
+    msg = "If eval_X is a tuple, y_val must be a tuple of same length, and vice versa."
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_X=(X_test, X_test), eval_y=y_test)
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_X=X_test, eval_y=(y_test, y_test))
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X_train, y_train, eval_X=(X_test,) * 3, eval_y=(y_test,) * 2)
+
+
+def test_ranker_eval_set_raises():
+    """Test that LGBMRanker raises expected errors from validating eval_group."""
+    X, y, g = _create_data(task="ranking", n_samples=1_000)
+    X_test, y_test, g_test = _create_data(task="ranking", n_samples=100)
+    gbm = lgb.LGBMRanker()
+
+    msg = "eval_group cannot be None if any of eval_set, eval_X, or eval_y are provided"
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_set=[(X_test, y_test)])
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_X=(X_test,))
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_y=(y_test,))
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_X=(X_test,), eval_y=(y_test,))
+
+    msg = re.escape("Length of eval_group (1) not equal to length of eval_set (2)")
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_X=(X_test, X_test), eval_y=(y_test, y_test), eval_group=[g_test])
+    with pytest.raises(ValueError, match=msg):
+        gbm.fit(X, y, group=g, eval_set=[(X_test, y_test), (X_test, y_test)], eval_group=[g_test])
+
+
+def test_eval_X_eval_y_eval_set_equivalence():
+    """Test that eval_X and eval_y are equivalent to eval_set."""
+    X, y = make_synthetic_regression()
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
+    params = {
+        "deterministic": True,
+        "force_row_wise": True,
+        "n_jobs": 1,
+        "seed": 708,
+    }
+    cbs = [lgb.early_stopping(2)]
+    gbm1 = lgb.LGBMRegressor(**params)
+    with pytest.warns(LGBMDeprecationWarning, match="The argument 'eval_set' is deprecated.*"):
+        gbm1.fit(X_train, y_train, eval_set=[(X_test, y_test)], callbacks=cbs)
+    gbm2 = lgb.LGBMRegressor(**params)
+    gbm2.fit(X_train, y_train, eval_X=X_test, eval_y=y_test, callbacks=cbs)
+    np.testing.assert_allclose(gbm1.predict(X), gbm2.predict(X))
+    assert gbm1.evals_result_["valid_0"]["l2"][0] == pytest.approx(gbm2.evals_result_["valid_0"]["l2"][0])
```

Comment:
_(changed by me in https://github.com/microsoft/LightGBM/pull/6857/commits/d7e0fff5cecc22e95027d43389667d857bdb48a3)_

Added this check on the validation results. Just checking the predicted values is not sufficient to test that the passed validation sets were actually used.